checksum.go

```diff
@@ -0,0 +1,213 @@
+package getter
+
+import (
+	"bufio"
+	"crypto/md5"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/hex"
+	"fmt"
+	"hash"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	urlhelper "github.com/hashicorp/go-getter/helper/url"
+)
+
+var checksummers = map[string]func() hash.Hash{
+	"md5":    md5.New,
+	"sha1":   sha1.New,
+	"sha256": sha256.New,
+	"sha512": sha512.New,
+}
+
+// checksumHashAndValue will return checksum based on checksum parameter of u
+// ex:
+//  http://hashicorp.com/terraform?checksum=<checksumType>:<checksumValue>
+//  http://hashicorp.com/terraform?checksum=file:<checksum_url>
+// when checksumming from a file checksumHashAndValue will go get checksum_url
+// in a temporary directory and parse the content of the file.
+// Content of files are expected to be BSD style or GNU style.
+//
+// BSD-style checksum:
+//  MD5 (file1) = <checksum>
+//  MD5 (file2) = <checksum>
+//
+// GNU-style:
+//  <checksum>  file1
+//  <checksum> *file2
+//
+// For GNU-style checksum files; it is very common that the hashing algorithm identifier
+// is in the filename; so the name of every supported hashing algorithm is compared
+// against checksum_url for a match/guess.
+// In case a different hashing algorithm is in the filename of checksum_url
+// it is recommended to explicitly set hashing algorithm instead.
+func checksumHashAndValue(u *url.URL) (checksumHash hash.Hash, checksumValue []byte, err error) {
+	q := u.Query()
+	v := q.Get("checksum")
+
+	if v == "" {
+		return nil, nil, nil
+	}
+
+	// Determine the checksum hash type
+	checksumType := ""
+	idx := len(v)
+	if i := strings.Index(v, ":"); i > -1 {
+		idx = i
+	}
+	checksumType = v[:idx]
+	if fn, found := checksummers[checksumType]; found {
+		checksumHash = fn()
+		// Get the remainder of the value and parse it into bytes
+		checksumValue, err = hex.DecodeString(v[idx+1:])
+		return
+	}
+
+	if checksumType != "file" {
+		return nil, nil, fmt.Errorf(
+			"unsupported checksum type: %s", checksumType)
+	}
+	file := v[idx+1:]
+
+	checkSums, err := checksumsFromFile(file, u)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var checksumValueString string
+	for checksumType, checksumValueString = range checkSums {
+		if fn, found := checksummers[checksumType]; found {
+			checksumHash = fn()
+			checksumValue, err = hex.DecodeString(checksumValueString)
+			return checksumHash, checksumValue, err
+		}
+	}
+
+	return nil, nil, fmt.Errorf(
+		"Could not find/guess checksum in %s: %s", file, checksumType)
+}
+
+// checksumsFromFile will download checksumFile that is a checksum for file
+// behind src.
+//
+// checksumsFromFile will try to guess the hashing algorithm based on content
+// of or name of checksum file
+func checksumsFromFile(checksumFile string, src *url.URL) (checkSums map[string]string, err error) {
+
```

Comment:
_Minor_ blank line here should be removed

Reply:
```suggestion
```